File: flutter_client/lib/services/api_client.dart
// ABOUTME: Client for communicating with the API server.
// ABOUTME: Handles SSE streaming for real-time agent responses.

import 'dart:async';
import 'dart:convert';
import 'package:http/http.dart' as http;
import 'package:logger/logger.dart';
import '../models/agent_event.dart';

class ApiClient {
  final String baseUrl;
  final http.Client _httpClient;
  final Logger _logger;

  ApiClient({
    required this.baseUrl,
    http.Client? httpClient,
    Logger? logger,
  })  : _httpClient = httpClient ?? http.Client(),
        _logger = logger ?? Logger();

  /// Sends a message and streams agent events.
  Stream<AgentEvent> sendMessage({
    required String conversationId,
    required String message,
  }) async* {
    final request = http.Request(
      'POST',
      Uri.parse('$baseUrl/chat'),
    );

    request.headers['Content-Type'] = 'application/json';
    request.body = jsonEncode({
      'conversationId': conversationId,
      'message': message,
    });

    final streamedResponse = await _httpClient.send(request);

    if (streamedResponse.statusCode != 200) {
      final body = await streamedResponse.stream.bytesToString();
      throw ApiException('Failed to send message: $body');
    }

    // Parse SSE stream
    String buffer = '';
    String? data;

    await for (final chunk in streamedResponse.stream.transform(utf8.decoder)) {
      buffer += chunk;
      final lines = buffer.split('\n');
      buffer = lines.last;

      for (int i = 0; i < lines.length - 1; i++) {
        final line = lines[i];

        if (line.startsWith('data: ')) {
          data = line.substring(6).trim();
          _logger.d('SSE data: $data');
        } else if (line.isEmpty && data != null) {
          try {
            final event = AgentEvent.fromSSEData(data);
            _logger.d('Parsed agent event: ${event.runtimeType}');
            yield event;
          } catch (e) {
            _logger.e('Error parsing agent event: $e');
          }
          data = null;
        }
      }
    }

    // Handle any remaining data
    if (data != null) {
      try {
        yield AgentEvent.fromSSEData(data);
      } catch (e) {
        _logger.e('Error parsing final agent event: $e');
      }
    }
  }

  void dispose() {
    _httpClient.close();
  }
}

class ApiException implements Exception {
  final String message;
  ApiException(this.message);

  @override
  String toString() => 'ApiException: $message';
}

File: react_client/src/services/api_client.ts
// ABOUTME: Client for communicating with the API server.
// ABOUTME: Handles SSE streaming for real-time agent responses.

import { type AgentEvent, parseAgentEventFromSSE } from "../models";

export class ApiException extends Error {
  constructor(message: string) {
    super(message);
    this.name = "ApiException";
  }
}

export class ApiClient {
  private readonly baseUrl: string;

  constructor(baseUrl: string) {
    this.baseUrl = baseUrl;
  }

  /**
   * Sends a message and yields agent events as they arrive via SSE.
   */
  async *sendMessage(
    conversationId: string,
    message: string
  ): AsyncGenerator<AgentEvent> {
    const response = await fetch(`${this.baseUrl}/chat`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify({ conversationId, message }),
    });

    if (!response.ok) {
      const body = await response.text();
      throw new ApiException(`Failed to send message: ${body}`);
    }

    if (!response.body) {
      throw new ApiException("Response body is null");
    }

    const reader = response.body.getReader();
    const decoder = new TextDecoder();

    let buffer = "";
    let data: string | null = null;

    try {
      while (true) {
        const { done, value } = await reader.read();

        if (done) {
          break;
        }

        buffer += decoder.decode(value, { stream: true });
        const lines = buffer.split("\n");
        buffer = lines.pop() ?? "";

        for (const line of lines) {
          if (line.startsWith("data: ")) {
            data = line.substring(6).trim();
          } else if (line === "" && data !== null) {
            try {
              const event = parseAgentEventFromSSE(data);
              yield event;
            } catch (e) {
              console.error("Error parsing agent event:", e);
              yield {
                type: "error" as const,
                errorType: "parse_error",
                message: `Failed to parse event: ${
                  e instanceof Error ? e.message : String(e)
                }`,
              };
            }
            data = null;
          }
        }
      }

      // Handle any remaining data in buffer
      if (data !== null) {
        try {
          yield parseAgentEventFromSSE(data);
        } catch (e) {
          console.error("Error parsing final agent event:", e);
          yield {
            type: "error" as const,
            errorType: "parse_error",
            message: `Failed to parse event: ${
              e instanceof Error ? e.message : String(e)
            }`,
          };
        }
      }
    } finally {
      reader.releaseLock();
    }
  }
}

// Default API URL - empty string uses relative URLs (works with Vite proxy)
const API_SERVER_URL = import.meta.env.VITE_API_SERVER_URL ?? "";

export const apiClient = new ApiClient(API_SERVER_URL);
